qoqo/src/circuitdag.rs

```diff
@@ -0,0 +1,366 @@
+// Copyright © 2021 HQS Quantum Simulations GmbH. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the
+// License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing permissions and
+// limitations under the License.
+
+///! Module containing the CircuitDag class that represents the Directed Acyclic Graph (DAG)
+///! of a quantum circuit in qoqo.
+///!
+use crate::{QoqoError, QOQO_VERSION};
+use bincode::{deserialize, serialize};
+use pyo3::exceptions::{PyIndexError, PyTypeError, PyValueError};
+use pyo3::prelude::*;
+use pyo3::types::{PyByteArray, PyType};
+use roqoqo::{Circuit, CircuitDag, ROQOQO_VERSION};
+
+use crate::operations::{convert_operation_to_pyobject, convert_pyany_to_operation};
+use crate::CircuitWrapper;
+
+/// Module containing the CircuitDag class that represents the Directed Acyclic Graph (DAG)
+/// of a quantum circuit in qoqo.
+///
+#[pymodule]
+fn circuitdag(_py: Python, module: &PyModule) -> PyResult<()> {
+    module.add_class::<CircuitDagWrapper>()?;
+    Ok(())
+}
+
+/// Represents the Direct Acyclic Graph (DAG) of a Circuit.
+///
+#[pyclass(name = "CircuitDag", module = "qoqo")]
+#[derive(Clone, Debug, PartialEq)]
+pub struct CircuitDagWrapper {
```

Comment:
We do not have it consistently in all of qoqo yet but at least for all new additions we should try to add functions signatures with PyO3's `function_signature` (https://pyo3.rs/v0.16.4/function.html). The signature for the new function goes here in fron of the struct definition. For other functions it goes in front of the function definition.